### ISSUE TITLE
Suggestion: External control of wrapping criteria

### ISSUE THREAD

**Issue by StylingAndroid** [google/flexbox-layout#32]
While it would certainly diverge from the CSS flexbox model upon which this library is based, there are some inherent inefficiencies for Android because of differences in cost in nesting `ViewGroup`s in Android compared to nesting `<div>` elements in HTML.

For tabular data, the following is simple and efficient in HTML:

```html
<div class="grid">
  <div class="grid__row">
    <div class="grid__item">1</div>
    <div class="grid__item">2</div>
  </div>
  <div class="grid__row">
    <div class="grid__item">1</div>
    <div class="grid__item">2</div>
    <div class="grid__item">3</div>
  </div>
  <div class="grid__row">
    <div class="grid__item">1</div>
    <div class="grid__item">2</div>
    <div class="grid__item">3</div>
    <div class="grid__item">4</div>
  </div>
</div>
```

And CSS flexbox can easily be applied to this. However for Android This would need to be done with a parent `ViewGroup` representing `grid`, containing an number for child `grid_row` `ViewGroup`s, each containing a collection of `View`s representing the individual the individual cells. Such a hierarchy is somewhat inefficient both in terms of layout inflation, and during measurement and layout.

A couple of simple additions to `FlexboxLayout` could make it easy to flatten such layouts by removing the individual rows - `FlexboxLayout` manages a collection of `FlexLine` objects which get calculated during measurement - and this is much more efficient that having child ViewGroups representing individual rows. If it were possible to have some optional attributes which could control when we should wrap to a new `FlexLine` then the usefulness of FlexboxLayout would increase dramatically, I think. Specifically we could have a `flexLineItems` integer which would cause a wrap whenever the current `FlexLine` contains the specified number of items. In addition a `LayoutParams.wrapAfter` boolean (default to `false`) could be used to force a wrap after the current child `View`.

As I said, I appreciate that this is a deviation from the CSS flexbox model, but feel that it could be a really useful addition on Android.

**Comment by thagikura** [comment]
Hi,

Thanks for your suggestion.
But I think having a explicit flag to indicate from which flex item the wrap should happen is not a way to go.

Because you don't have to nest ViewGroups to do so in the grid example above.
If you want to build a grid with the `FlexboxLayout` you can put the grid cells as direct children of the `FlexboxLayout`. 

E.g. If you want make the first row having two cells, the second row having three cells and the third row having four cells, you can write a XML like below (omitting the not relevant attributes).

```xml
<com.google.android.flexbox.Flexbox
  app:flexWrap="wrap">

    <View app:layout_flexBasisPercent="50%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="50%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="30%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="30%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="30%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="25%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="25%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="25%"
          app:layout_flexGrow="1">
    <View app:layout_flexBasisPercent="25%"
          app:layout_flexGrow="1">

</com.google.android.flexbox.Flexbox>
```
I think the layout above gives you the same result as you HTML as you pasted.

Instead if you want to give each flex item a fixed initial width and want to make the FlexboxLayout grow (have a new flex line) when a flex item doesn't fit in the current flex line, you can also do so like the example in the [wiki](https://github.com/google/flexbox-layout/wiki/Target-use-cases) 

Having that flexibility is one of the big advantages of the Flexbox and I think adding a explicit flag to indicate from which flex item the wrap should happen loses that flexibility.

How do you think?

**Comment by StylingAndroid** [comment]
But that only works if you manually specify the percentages and can already be achieve efficiently on Android using `PercentLayout`. Having the ability to control when wraps occur would make `FlexboxLayout` totally unique for Android, and would open up many more use-cases which only `FlexboxLayout` would provide an efficient solution.

**Comment by thagikura** [comment]
Ok, I understand controlling when wrap occurs is equivalent to having fixed number of items in each grid row.
So I think it can be achieved by having different span counts on each row in a grid using GridLayoutManager like [this](http://stackoverflow.com/questions/31112291/recyclerview-layoutmanager-different-span-counts-on-different-rows).

**Comment by StylingAndroid** [comment]
That would work, but would further compilicate an API which already isn't that intuitive for Android devs (alignItems vs. alignContent). That's the primary reason I'm suggesting what I am - increased control over internal behaviour via simple, understandable API additions.

However, any additions to the functionality which would facilitate flattened layouts for tabular data would be greatly welcomed by me ;-)

**Comment by thagikura** [comment]
Sorry for the long delay.
During the course of replacing the existing LinearLayout with FlexboxLayout ( #59 ), I realized that attribute is useful in many situations.

I'm going to add this request to the queue of the upcoming implementation plan.
Thanks.

**Comment by thagikura** [comment]
Fixed by #60 
I'm going to include it in the next release.
